src/vga.js

```diff
@@ -409,7 +408,66 @@ function VGAScreen(cpu, bus, vga_memory_size)
         function(addr, value) { me.vga_memory_write(addr, value); }
     );
 
+    if(!! screen_options?.use_graphical_text)
+    {
+        this.graphical_text = new GraphicalText(bus, me);
+    }
+    this.text_to_graphical_mode_flag = false;   // flag that indicates a previous text-to-graphical mode transition to screen_fill_buffer()
+
     cpu.devices.pci.register_device(this);
+
+    setTimeout(() => { me.switch_screen_mode(this.graphical_mode); }, 0);
+}
+
+/* function switch_screen_mode(): 3-state screen mode: text, graphical-text or graphical
+ * - when transitioning from text to graphical mode, set flag for screen_fill_buffer() to call update_vertical_retrace() once
+ * - in graphical-text mode only report graphical mode to ScreenAdapter but actually work in text mode
+ * - restore canvas size when switching to either graphical or graphical-text mode
+ * - single function that sends bus event "screen-set-mode" to ScreenAdapter
+ */
+VGAScreen.prototype.switch_screen_mode = function(graphical_mode)
+{
+    const previous_graphical_mode = this.graphical_mode;
+    this.graphical_mode = graphical_mode;
+
+    if(!previous_graphical_mode && graphical_mode)
+    {
+        this.text_to_graphical_mode_flag = true;
+    }
+
+    if(!graphical_mode && this.graphical_text)
+    {
+        this.graphical_text.resize_canvas();
+    }
+
+    this.bus.send("screen-set-mode", graphical_mode || this.graphical_text);
+}
+
+VGAScreen.prototype.grab_text_content = function(keep_whitespace)
```

Comment:
This looks unused.

Reply:
Early on you mentioned the current ability to select text from the DIV using the mouse.

This is my poor-man's replacement for this, but I left out the accompanying method in V86's user interface, currently I call it using v86.cpu.devices.vga.grab_text_content() in my V86 embedding.

I could add mouse-controlled text-selection though, roughly like this:

- Text selection is simply a linear copy from the 16 bit VGA text buffer, starting at (start-row, start-col) to (end-row, end-col).
- If any character within the current selection is changed, selection gets cancelled and reset to nothing.
- Swap forground and background color when rendering selected text.
- Immediately copy to clipboard when user changes selected text range.
- Any events from mouse and keyboard related to text selection must be hidden from the virtualized OS. Not sure yet how to modify MouseAdapter and KeyboardAdapter here.

But someone would have to establish a mouse-button + key combination that triggers text selection in graphical text mode. For example, I use the DOS mouse driver in FreeDOS, and I can select/copy text *within* FreeDOS's EDIT.EXE using my mouse. So the normal left mouse button is not an option, text-based applications may use it in graphical text mode. IIRC already DOS supported mice with 3 buttons, and certainly the 104/105 keyboards.

Or, make text selection in graphical text mode a special state that has to be triggered through the HTML user interface, naively like a button "Begin Text Selection" which changes its label to "End Text Selection" when clicked, and during that state the left mouse button would be used for text selection normally and all mouse events would be hidden from the virtualized OS. I'd prefer that.

What do you think?

Reply:
> I could add mouse-controlled text-selection though, roughly like this:

This sounds reasonable. I would change the following:

> If any character within the current selection is changed, selection gets cancelled and reset to nothing.

In my experience, this makes copying regularly changing text annoying, and the alternative is better. 

> Immediately copy to clipboard when user changes selected text range.

I don't think this is a good idea, as users expect the "browser-like" behaviour of copying with ctrl+c. It's also likely that browser's don't allow changing the clipboard without user interaction.

An alternative implementation could use regular DOM text elements with the current vga text implementation, that are "above" the canvas screen but opaque. I believe xterm.js does something like that, but I can't find the implementation right now.

> Or, make text selection in graphical text mode a special state that has to be triggered through the HTML user interface, naively like a button "Begin Text Selection" which changes its label to "End Text Selection" when clicked, and during that state the left mouse button would be used for text selection normally and all mouse events would be hidden from the virtualized OS. I'd prefer that.

That sounds reasonable (much better than a mouse-button + key combination), and it's also similar how text selection in the existing text screen works: When the guest OS uses the mouse, clicking on the screen doesn't select text, it locks the mouse. When the user clicks on "disable mouse" (or the guest OS doesn't use the mouse in the first place), text selection is provided by the browser.

Reply:
Fully agree (explicit select/copy-state and Ctrl+C to copy).

The canvas overlay idea is good and I will look into it when I find time. Maybe one way to go would be to use [context.fillRect()](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/fillRect) with [context.globalCompositeOperation = "xor"](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/globalCompositeOperation) to invert text in the browser canvas at low costs (as oppsed to the render() function of GraphicalText).

A HTML DIV overlay in graphical text mode would make the font appearance mismatch. It also touches on the issue of the dreaded 8-bit codepages and their Unicode-mappings.

However, I'd vote to leave this issue for another day.

Reply:
See this [little demo](https://www.rekoba.de/~cs/v86/select-demo/) about highlighting selected text in a canvas using xor. That on the top is a canvas with a static image, the raw text data is simply stored fixed in the js file.

context.globalCompositeOperation isn't what it seems.